{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Session 2: Data normalization with scVI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Google colab setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "if 'google.colab' in sys.modules:\n",
    "    !pip install --quiet scrnatools\n",
    "    !pip install --quiet scvi-colab\n",
    "    from scvi_colab import install\n",
    "    install()\n",
    "    from google.colab import drive\n",
    "    drive.mount(\"/content/gdrive\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n"
     ]
    }
   ],
   "source": [
    "import scrnatools as rna\n",
    "import scanpy as sc\n",
    "import os\n",
    "import shutil\n",
    "import scvi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"data/scRNAseq_tutorial_thymus_data.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup scVI model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To train an scVI model, you first need to set up the variables you want the network to use wile learning. The main parameters to adjust are th covariates you want the model to be aware of while 'learning' a particular cell's gene expression. These include:\n",
    "- Categorical covariates: these should be biologically irrelevant sample-specific nuisance variables that you want to correct for, i.e. biological replicates\n",
    "- Continuous covariates: these are similar to categorical nuisance covariates, except that they take on continuous values (i.e. total counts per cell, mitochonrial read %)\n",
    "- A batch key: this is the batch categorical covariate you want to use to integrate diffferent *biologically relevant* samples (i.e. WT vs KO samples)\n",
    "\n",
    "The important distinction to make between categorical covariates and the batch key is that the batch key encodes samples you want to integrate but *expect to have meaningful biological differences*. In practce this means that by default the batch key and categorical covariates will both be accounted for/corrected in the scVI latent space and downstream dimensionality reductions (UMAPs) but only catgorical covariates' (not batch key's) effects will be removed from the normalized per-cell gene expression that is generated by scVI as well as downstream differential gene expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)\n"
     ]
    }
   ],
   "source": [
    "# Setup scVI model using raw counts and batch info\n",
    "scvi.model.SCVI.setup_anndata(\n",
    "    adata,\n",
    "    layer=\"raw_counts\", # scVI models should always be trained on true raw counts (not log-transformed)\n",
    "    categorical_covariate_keys=None,\n",
    "    continuous_covariate_keys=None,\n",
    "    batch_key=\"genotype\", # The batch\n",
    ")\n",
    "model = scvi.model.SCVI(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train scVI model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Training the scVI model will take a very long time unless it is run on a machine with a discrete GPU (i.e. not most laptops). It will also take longer the more genes and cells are in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "IPU available: False, using: 0 IPUs\n",
      "HPU available: False, using: 0 HPUs\n",
      "You are using a CUDA device ('NVIDIA GeForce RTX 3090') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 400/400: 100%|██████████| 400/400 [03:55<00:00,  1.79it/s, loss=1.43e+03, v_num=1]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`Trainer.fit` stopped: `max_epochs=400` reached.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 400/400: 100%|██████████| 400/400 [03:55<00:00,  1.70it/s, loss=1.43e+03, v_num=1]\n"
     ]
    }
   ],
   "source": [
    "# Train scVI model - run on google colab with a GPU session or a personal computer/server with an Nvidia GPU\n",
    "model.train()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get scVI normalized expression/latent representation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The scVI model learns a 15-dimension latent representation of the expression data that is then used for downstream clustering and further dimensionality reduction (i.e. UMAP/TSNE). This latent representation is learned to correct all batch effects encoded by the batch_key, continuous_covariates, and categorical_covariates supplied during model setup."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm[\"X_scVI\"] = model.get_latent_representation()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "scVI normalized gene expression learned during model training is encoded as a non-zero probability that corresponds to the denoised expression frequency in a given cell. This is scaled by the provided library size (usually 10,000, but this is arbitrary) to give an estimate on the number of reads of a particular gene per 10,000 total reads expected to be observed in the given cell. This normalized expression is generated based on a cell's original 'batch_key' by default, which is to say that any biological differences in gene expression across different 'batch_keys' is preserved, while differences in gene expression across different 'categorical_covariates' are lost. This means that a gene that has variability in expression acrosss WT and KO samples maintains that variability, while expression that varies across biological replicates is removed (as it iss likely due to technical noise)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers[\"scVI_normalized\"] = model.get_normalized_expression(library_size=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving/Loading scVI models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "scVI model training is non-deterministic, so it is important to save your model immediately to allow downstream analyses to be repeatable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If a model folder with the specified path/name already exists you have to delete it before resaving the scVI model\n",
    "if os.path.isdir(\"data/scRNAseq_tutorial_thymus_data_scVI_model\"):\n",
    "    shutil.rmtree(\"data/scRNAseq_tutorial_thymus_data_scVI_model\")\n",
    "    \n",
    "model.save(\"data/scRNAseq_tutorial_thymus_data_scVI_model\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m File data/scRNAseq_tutorial_thymus_data_scVI_model/model.pt already downloaded                            \n"
     ]
    }
   ],
   "source": [
    "model = scvi.model.SCVI.load(\"data/scRNAseq_tutorial_thymus_data_scVI_model\", adata=adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"data/scRNAseq_tutorial_thymus_data.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "single-cell-analysis",
   "language": "python",
   "name": "single-cell-analysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
